api/classes/publisher.js

```diff
@@ -79,72 +118,238 @@ function deleteFile(file, root) {
   });
 }
 
-function modifyFiles(action, project, user) {
-  return File.query({
-    where: {
-      project_id: project.id
+function modifyPublishedFiles(action, project, user, files) {
+  return Promise.resolve()
+  .then(function() {
+    if (files.length) {
+      return files.mapThen(function(file) {
+        return action(file, '/' + user.get('name') + '/' + project.get('id'));
+      });
+    } else {
+      return action(files, '/' + user.get('name') + '/' + project.get('id'));
     }
-  }).fetchAll()
-  .then(function(records) {
-    return records.mapThen(function(record) {
-      return action(record, '/' + user.get('name') + '/' + project.get('id'));
+  });
+}
+
+/**
+ * Exports
+ */
+exports.publish = function publish(project) {
+  var user;
+  var files;
+
+  var publishedFiles;
+  var publishedProject;
+
+  var projectFiles = {};
+  var publishedProjectFiles = {};
+  var toBeCreated = [];
+  var toBeDeleted = [];
+  var toBeUpdated = [];
+
+  return Promise.resolve()
+  .then(function() {
+    return getUserForProject(project)
+    .then(function(u) {
+      user = u;
     });
   })
   .then(function() {
-    return buildUrl(project, user);
+    return getPublishedProject(project)
+    .then(function(oldPublishedProject) {
+      if (oldPublishedProject) {
+        return oldPublishedProject.set({
+          title: project.get('title'),
+          tags: project.get('tags'),
+          description: project.get('description')
+        }).save();
+      } else {
+        return PublishedProject.forge({
+          title: project.get('title'),
+          tags: project.get('tags'),
+          description: project.get('description')
+        }).save();
+      }
+    })
+    .then(function(p) {
+      publishedProject = p;
+    });
   })
-  .catch(function(e) {
-    return Promise.reject(e);
-  });
-}
+  .then(function() {
+    return File.query({
+      where: {
+        project_id: project.get('id')
+      }
+    }).fetchAll()
+    .then(function(originalFiles) {
+      if (originalFiles.length === 0) {
+        throw 'No files to be published!';
+      }
 
-function getUser(project) {
-  return project.user().query({}).fetch();
-}
+      files = originalFiles;
+    });
+  })
+  .then(function() {
+    return PublishedFiles.query({
+      where: {
+        published_id: publishedProject.get('id')
+      }
+    }).fetchAll()
+    .then(function(f) {
+      if (f.length === 0) {
+        return;
+      }
+      publishedFiles = f;
+    });
+  })
+  .then(function() {
+    // We want to organize 'file' objects by id, and
+    // 'publishedFile' objects by the 'file' objects they reference
+    files.models.forEach(function(file) {
```

Comment:
Consider using this instead: http://bookshelfjs.org/#Collection-Underscore-Methods